### PR TITLE
Vss over Can implementation proposal

### PR DESCRIPTION
**Pull request by shadowm60**
Working for BMW e46.

Can i get a review?

### CODE REVIEW

**Comment by rusefi** [comment]
Can one of the admins verify this patch? (meaning the author is not on the white list)

**Comment by rusefi** [comment]
Can one of the admins verify this patch? (meaning the author is not on the white list)

**Comment by rusefi** [comment]
Can one of the admins verify this patch? (meaning the author is not on the white list)

**Comment by rusefi** [comment]
Can one of the admins verify this patch? (meaning the author is not on the white list)

File: firmware/hw_layer/vehicle_speed.cpp [hunk 1]
```diff
@@ -92,11 +92,6 @@ void initVehicleSpeed(Logging *l) {
 #else  /* EFI_VEHICLE_SPEED */
 
 float getVehicleSpeed(void) {
-#if EFI_CAN_SUPPORT
-	if ( CONFIG(enableCanVss) == true ) {
-		return getVehicleCanSpeed();
-	}
-#endif
```
Comment (rusefi) [inline firmware/hw_layer/vehicle_speed.cpp]: our code style is towards
``#endif /* EFI_CAN_SUPPORT */``

**Comment by rusefi** [comment]
kinetis does not build - need more compiler protections, see error failure and error message

File: firmware/controllers/can/can_vss.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    if ( (nowNt - frameTime ) > NT_PER_SECOND ) {
+        return 0; /* can timeout? */
+    } else {
+        return vssSpeed;
+    }
+}
+
+void initCanVssSupport(Logging *l) {
+
+    if( CONFIG(enableCanVss) == true ) {
```
Comment (rusefi) [inline firmware/controllers/can/can_vss.cpp:94]: space BEFORE bracket and NO space AFTER opening bracket

File: firmware/controllers/can/can_vss.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    efitick_t nowNt = getTimeNowNt();
+
+    if ( (nowNt - frameTime ) > NT_PER_SECOND ) {
+        return 0; /* can timeout? */
+    } else {
+        return vssSpeed;
+    }
+}
+
+void initCanVssSupport(Logging *l) {
```
Comment (rusefi) [inline firmware/controllers/can/can_vss.cpp:92]: "l" parameter name is a bit laconic. "logging" maybe?

File: firmware/controllers/can/can_vss.cpp (hunk 2 above)
Comment (rusefi) [reply firmware/controllers/can/can_vss.cpp:94]: if enableCanVss is a bit we do not need "== true" part

**Comment by rusefi** [comment]
this is pretty mergable as is, just fix kinetis :)

**Comment by shadowm60** [comment]
> this is pretty mergable as is, just fix kinetis :)

i am trying to fix that as well, and then the small findings as well, just needed an ok, if it is ok to continue or not.

**Comment by shadowm60** [comment]
For the w202 i have to take traces from my car since i did not found the can message content. so that is next step.

**Comment by rusefi** [comment]
This is a great PR, will merge later today

**Comment by rusefi** [comment]
add to white list

**Comment by rusefi** [comment]
add to whitelist

**Pull request merged**